File: 13.Linked_List/javascript/11.NthNodeFromEnd.js
class Node{
    constructor(data){
        this.data = data
        this.next = null
    }
}

function buildLinkedList(arr){
    
    let head = new Node(arr[0])
    let current = head
    
    for(let i=1;i<arr.length;i++){
        current.next = new Node(arr[i])
        current = current.next
    }
    
    return head
}


class Solution{
    getNthFromLast(head,n) {
        
        let first = head
        let second = head
        
        let current = head
        
        for(let i=0;i<n;i++){
            first = current.next 
        }
        
        while(first !==null){
            first = first.next
            second = second.next
        }
        
        return second ? second : null
        
    }
}


function  PrintLinkedList(head){
    let res = ''
    let current = head
    
    while(current !==null){
        res += current.data + ' '
        current = current.next
    }
    
    console.log(res)
}


let arr = [1,2,3,4,5]
let Sol = new Solution()
let n = 1
let head = buildLinkedList(arr)
let newNode = Sol.getNthFromLast(head,n)
PrintLinkedList(newNode)

File: 13.Linked_List/python/11.NthNodeFromEnd.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def getNthFromLast(head, n):
    first = head
    second = head
    
    # Move first pointer n steps ahead
    for _ in range(n):
        if first is None:
            return -1  # n is greater than the length of the list
        first = first.next
    
    # Move both pointers until first reaches the end
    while first:
        first = first.next
        second = second.next

    return second.data

# Constructing the linked list: 1 -> 2 -> 3
head = Node(1)
head.next = Node(2)
head.next.next = Node(3)

n = 1
print(getNthFromLast(head, n))  # Output: 3
